File: neuron_db/plastic.py
"""PlasticNeuron -- the store tier of a plastic memory: a neuron whose recall changes
with use, forms associations, and forgets, using only cheap O(1) scalar updates. No
neural net runs here; this is the substrate that decides WHAT enters the working set.

The thinking tier (the trained gary-neuron cortex + plastic hippocampus) runs on top of
the small working set this returns -- see docs/PLASTICITY.md. Plasticity is split so the
expensive neural part only ever sees a bounded window, never the whole store.

Mechanisms (all O(1) or O(neighbors), no background sweeps):
  * strength       each fact carries a weight, bumped on recall (Hebbian "use it")
  * decay          weight = w * 0.5 ** (age / half_life), computed LAZILY at read time
  * association    facts recalled close together get linked (fire together, wire together)
  * spreading      recall can return the hit PLUS its strongest associates (one hop)
  * consolidate    off-hot-path: merge duplicate-stem facts, prune decayed-to-nothing
"""
from __future__ import annotations
from typing import Optional
from .neuron import (Neuron, _content, _stem, _pick_value, REL_S, PETS, STOPVAL_S, _encode, ENT, LEADIN, _words, COMMAND, _sents)


class PlasticNeuron(Neuron):
    def __init__(self, max_facts: int = 500, half_life: float = 200.0, link_window: int = 3):
        super().__init__(max_facts)
        self.half_life = half_life          # in ticks; smaller = forgets faster
        self.link_window = link_window      # how many recent recalls a new recall links back to
        self.tick = 0
        self._w: dict[int, float] = {}      # episode id -> strength
        self._t: dict[int, int] = {}        # episode id -> last-touched tick
        self._links: dict[int, dict[int, float]] = {}  # id -> {id: weight}
        self._recent: list[int] = []        # recently activated ids (for co-activation linking)
        self._next_id = 0

    # ----- write -----
    def observe(self, text: str, entity: Optional[str] = None) -> list:
        wrote = super().observe(text, entity)   # returns the newly-appended episode dicts
        new_ids = []
        for e in wrote:                          # same dict objects that live in self.episodes
            e["_id"] = self._next_id; self._next_id += 1
            self._w[e["_id"]] = 1.0; self._t[e["_id"]] = self.tick
            new_ids.append(e["_id"])
        # facts stated together are structurally associated
        for i in range(len(new_ids)):
            for j in range(i + 1, len(new_ids)):
                self._link(new_ids[i], new_ids[j], 1.0)
        self.tick += 1
        return wrote

    # ----- plastic helpers -----
    def _eff(self, eid: int) -> float:
        age = self.tick - self._t.get(eid, self.tick)
        return self._w.get(eid, 1.0) * (0.5 ** (age / self.half_life))

    def _link(self, a: int, b: int, d: float):
        if a == b: return
        self._links.setdefault(a, {})[b] = self._links.setdefault(a, {}).get(b, 0.0) + d
        self._links.setdefault(b, {})[a] = self._links.setdefault(b, {}).get(a, 0.0) + d

    def reinforce(self, eid: int, amount: float = 1.0):
        self._w[eid] = self._eff(eid) + amount; self._t[eid] = self.tick

    # ----- read: same matching as Neuron, but ties broken by effective strength -----
    def recall(self, query) -> Optional[dict]:
        cue = _stem(_content(query)) if isinstance(query, str) else _stem(set(query))
        if not cue: return None
        pet_query = bool(cue & _stem({"pet", "animal"}))
        name_query = ("name" in cue) and not (cue & REL_S)
        if self._index is None or self._index_len != len(self.episodes): self._build_index()
        cand: set = set()
        for s in cue: cand.update(self._index.get(s, ()))
        if pet_query:
            for s in PETS: cand.update(self._index.get(s, ()))
        best = None; bk = (-1, -1.0, -1, -1, 0, -1)
        for i in sorted(cand):
            e = self.episodes[i]
            es = set(e["s"]); ov = len(cue & es); es_pet = bool(es & PETS)
            if ov < 1 and pet_query and es_pet: ov = 1
            if ov < 1: continue
            if (es & REL_S) - cue and not (pet_query and es_pet): continue
            if (cue & REL_S) - es and not (pet_query and es_pet): continue
            eid = e.get("_id", -1)
            selfp = 1 if (name_query and e.get("self")) else 0
            spec = -len(es - cue - STOPVAL_S)
            # overlap first (a clearly better match still wins), THEN learned strength,
            # then the original tie-breakers. This is the adaptive bit: a frequently-used
            # fact beats a merely-recent one on an otherwise-equal cue.
            sc = (ov, self._eff(eid), selfp, 1 if e.get("h", "") in cue else 0, spec, i)
            if sc > bk: bk = sc; best = e
        if best is None: return None
        bes = set(best["s"]); cov = len(cue & bes) / max(1, len(cue))
        if pet_query and (bes & PETS): cov = 1.0
        val, echo = _pick_value(best, cue, bool(cue & _stem({"many", "much", "number"})))
        # plastic side effects: reinforce the hit, wire it to recently-activated facts
        eid = best.get("_id", -1)
        if eid >= 0:
            self.reinforce(eid)
            for prev in self._recent[-self.link_window:]:
                self._link(eid, prev, 0.5)
            self._recent.append(eid); self._recent = self._recent[-32:]
            self.tick += 1
        return {"fact": best["t"], "value": val, "coverage": cov, "overlap": bk[0], "echo": echo, "strength": round(self._eff(eid), 3)}

    def recall_related(self, query, k: int = 3) -> list:
        """The hit plus its strongest associates -- one-hop spreading activation."""
        hit = self.recall(query)
        if hit is None: return []
        out = [hit]
        eid = next((e["_id"] for e in self.episodes if e["t"] == hit["fact"]), -1)
        nbrs = sorted(self._links.get(eid, {}).items(), key=lambda kv: -kv[1] * self._eff(kv[0]))
        by_id = {e.get("_id"): e for e in self.episodes}
        for nid, w in nbrs[:k]:
            e = by_id.get(nid)
            if e: out.append({"fact": e["t"], "value": e["v"], "link": round(w, 2), "strength": round(self._eff(nid), 3)})
        return out

    # ----- consolidation ("sleep"): off the hot path -----
    def consolidate(self, prune_below: float = 0.05) -> dict:
        merged = 0; pruned = 0
        # merge facts with identical stem-sets, keeping the strongest, summing strength
        by_key: dict[tuple, dict] = {}
        keep = []
        for e in self.episodes:
            key = tuple(e["s"])
            if key in by_key:
                a = by_key[key]; ai = a["_id"]; bi = e["_id"]
                self._w[ai] = self._eff(ai) + self._eff(bi); self._t[ai] = self.tick
                for nid, w in self._links.pop(bi, {}).items():
                    self._link(ai, nid, w)
                merged += 1
            else:
                by_key[key] = e; keep.append(e)
        # prune facts whose effective strength has decayed away
        survivors = []
        for e in keep:
            if self._eff(e["_id"]) >= prune_below or e.get("self"):
                survivors.append(e)
            else:
                pruned += 1; self._w.pop(e["_id"], None); self._t.pop(e["_id"], None); self._links.pop(e["_id"], None)
        self.episodes = survivors
        self._index = None
        return {"merged": merged, "pruned": pruned, "facts": len(self.episodes)}

File: rust/neuron-core/src/plastic.rs
//! PlasticNeuron: the adaptive store tier in pure Rust. A Neuron whose recall changes with
//! use -- strength on recall, lazy exponential decay, Hebbian links, and a neurotransmitter
//! style spreading-activation recall. All O(1) / O(neighbors) scalar updates, no neural net.
//! Faithful port of neuron_db/plastic.py.

use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use super::{Neuron, Recall, content, stems_s, stem1, rel_s, pets, stopval_s, pick_value, has_stem};

/// One readout of the spreading-activation pass.
#[derive(Clone, Debug)]
pub struct Spread { pub fact: String, pub value: String, pub activation: f64, pub seed: bool }

pub struct PlasticNeuron {
    pub base: Neuron,
    pub half_life: Option<f64>,   // None / <=0 / inf => no decay
    pub link_window: usize,
    pub tick: i64,
    w: HashMap<i64, f64>,
    t: HashMap<i64, i64>,
    links: HashMap<i64, HashMap<i64, f64>>,
    recent: Vec<i64>,
    pinned: HashSet<i64>,
    next_id: i64,
}

#[inline]
fn gt(a: &(i64, f64, i64, i64, i64, i64), b: &(i64, f64, i64, i64, i64, i64)) -> bool {
    if a.0 != b.0 { return a.0 > b.0; }
    if a.1 != b.1 { return a.1 > b.1; }
    if a.2 != b.2 { return a.2 > b.2; }
    if a.3 != b.3 { return a.3 > b.3; }
    if a.4 != b.4 { return a.4 > b.4; }
    a.5 > b.5
}

impl PlasticNeuron {
    pub fn new(max_facts: usize, half_life: Option<f64>, link_window: usize) -> Self {
        PlasticNeuron { base: Neuron::new(max_facts), half_life, link_window, tick: 0,
            w: HashMap::new(), t: HashMap::new(), links: HashMap::new(),
            recent: Vec::new(), pinned: HashSet::new(), next_id: 0 }
    }
    pub fn fact_count(&self) -> usize { self.base.fact_count() }
    pub fn strength(&self, id: i64) -> f64 { self.eff(id) }

    fn eff(&self, id: i64) -> f64 {
        let w = *self.w.get(&id).unwrap_or(&1.0);
        match self.half_life {
            None => w,
            Some(h) if h <= 0.0 || h.is_infinite() => w,
            Some(h) => { let age = (self.tick - *self.t.get(&id).unwrap_or(&self.tick)) as f64; w * 0.5f64.powf(age / h) }
        }
    }
    pub fn pin(&mut self, id: i64) { self.pinned.insert(id); }
    fn link(&mut self, a: i64, b: i64, d: f64) {
        if a == b { return; }
        *self.links.entry(a).or_default().entry(b).or_insert(0.0) += d;
        *self.links.entry(b).or_default().entry(a).or_insert(0.0) += d;
    }
    pub fn reinforce(&mut self, id: i64, amount: f64) {
        let e = self.eff(id); self.w.insert(id, e + amount); self.t.insert(id, self.tick);
    }

    pub fn observe(&mut self, text: &str) -> usize {
        let before = self.base.episodes.len();
        let n = self.base.observe(text);
        // PlasticNeuron is meant to run with a large max_facts (no drain); the new episodes
        // are the last `n` appended.
        let mut ids = Vec::new();
        let len = self.base.episodes.len();
        if before + n <= len {
            for k in 0..n {
                let idx = before + k;
                let id = self.next_id; self.next_id += 1;
                self.base.episodes[idx].id = id;
                self.w.insert(id, 1.0); self.t.insert(id, self.tick);
                ids.push(id);
            }
        }
        for i in 0..ids.len() { for j in i + 1..ids.len() { self.link(ids[i], ids[j], 1.0); } }
        self.tick += 1;
        n
    }

    /// Same matching as Neuron::recall, but ties break on learned strength, and the hit is
    /// reinforced and wired to recently-recalled facts (the adaptive side effects).
    pub fn recall(&mut self, query: &str) -> Option<Recall> {
        let cue: HashSet<String> = stems_s(&content(query));
        if cue.is_empty() { return None; }
        let qraw: HashSet<String> = content(query);
        let pet_query = cue.contains(&stem1("pet")) || cue.contains(&stem1("animal"));
        let name_query = cue.contains("name") && cue.intersection(rel_s()).count() == 0;
        let order = self.base.candidates(&cue, pet_query);
        let mut best: Option<usize> = None;
        let mut bk = (-1i64, -1f64, -1i64, -1i64, -1i64, -1i64);
        for i in order {
            let e = &self.base.episodes[i];
            let mut ov = cue.iter().filter(|c| has_stem(&e.s, c)).count();
            let es_pet = e.s.iter().any(|s| pets().contains(s.as_ref()));
            if ov < 1 && pet_query && es_pet { ov = 1; }
            if ov < 1 { continue; }
            if e.s.iter().any(|s| rel_s().contains(s.as_ref()) && !cue.contains(s.as_ref())) && !(pet_query && es_pet) { continue; }
            if cue.iter().any(|s| rel_s().contains(s) && !has_stem(&e.s, s)) && !(pet_query && es_pet) { continue; }
            let exact_ov = qraw.iter().filter(|wd| has_stem(&e.raw, wd)).count() as i64;
            let selfp = if name_query && e.self_flag { 1 } else { 0 };
            let subj = if cue.contains(&e.head) { 1 } else { 0 };
            let spec = -(e.s.iter().filter(|s| !cue.contains(s.as_ref()) && !stopval_s().contains(s.as_ref())).count() as i64);
            let sc = (exact_ov, self.eff(e.id), selfp, subj, spec, i as i64);
            if gt(&sc, &bk) { bk = sc; best = Some(i); }
        }
        let bi = best?;
        let (fact, sset, id) = { let e = &self.base.episodes[bi]; (e.t.clone(), e.s.clone(), e.id) };
        let mut cov = cue.iter().filter(|c| has_stem(&sset, c)).count() as f64 / (cue.len().max(1) as f64);
        if pet_query && sset.iter().any(|s| pets().contains(s.as_ref())) { cov = 1.0; }
        let overlap = cue.iter().filter(|c| has_stem(&sset, c)).count();
        let exact_n = bk.0 as usize;
        let want_num = cue.contains("many") || cue.contains("much") || cue.contains(&stem1("number"));
        let (val, echo) = pick_value(&self.base.episodes[bi], &cue, want_num);
        if id >= 0 {
            self.reinforce(id, 1.0);
            let recents: Vec<i64> = self.recent.iter().rev().take(self.link_window).cloned().collect();
            for prev in recents { self.link(id, prev, 0.5); }
            self.recent.push(id);
            let l = self.recent.len(); if l > 32 { self.recent.drain(0..l - 32); }
            self.tick += 1;
        }
        Some(Recall { fact, value: val, coverage: cov, overlap, exact: exact_n, echo, idx: bi })
    }

    /// The hit plus its strongest associates (one-hop spreading).
    pub fn recall_related(&mut self, query: &str, k: usize) -> Vec<(String, String, f64)> {
        let hit = match self.recall(query) { Some(h) => h, None => return Vec::new() };
        let eid = self.base.episodes.iter().find(|e| e.t == hit.fact).map(|e| e.id).unwrap_or(-1);
        let mut out = vec![(hit.fact.clone(), hit.value.clone(), 0.0f64)];
        let mut nbrs: Vec<(i64, f64)> = self.links.get(&eid).map(|m| m.iter().map(|(a, b)| (*a, *b)).collect()).unwrap_or_default();
        nbrs.sort_by(|a, b| (b.1 * self.eff(b.0)).partial_cmp(&(a.1 * self.eff(a.0))).unwrap());
        let by_id: HashMap<i64, usize> = self.base.episodes.iter().enumerate().map(|(i, e)| (e.id, i)).collect();
        for (nid, w) in nbrs.into_iter().take(k) {
            if let Some(&ix) = by_id.get(&nid) { let e = &self.base.episodes[ix]; out.push((e.t.clone(), e.v.clone(), w)); }
        }
        out
    }

    /// Neurotransmitter-style recall: excitatory release at cue matches, inhibitory relation
    /// gating, multi-hop spread across synapses with reuptake decay. Pure read (no mutation).
    pub fn recall_spreading(&mut self, query: &str, hops: usize, k: usize, reuptake: f64, fan: usize) -> Vec<Spread> {
        let cue: HashSet<String> = stems_s(&content(query));
        if cue.is_empty() { return Vec::new(); }
        let pet_query = cue.contains(&stem1("pet")) || cue.contains(&stem1("animal"));
        let order = self.base.candidates(&cue, pet_query);
        let by_id: HashMap<i64, usize> = self.base.episodes.iter().enumerate().map(|(i, e)| (e.id, i)).collect();
        let mut act: HashMap<i64, f64> = HashMap::new();
        for i in order {
            let e = &self.base.episodes[i];
            let mut ov = cue.iter().filter(|c| has_stem(&e.s, c)).count();
            let es_pet = e.s.iter().any(|s| pets().contains(s.as_ref()));
            if ov < 1 && pet_query && es_pet { ov = 1; }
            if ov < 1 { continue; }
            if e.s.iter().any(|s| rel_s().contains(s.as_ref()) && !cue.contains(s.as_ref())) && !(pet_query && es_pet) { continue; }
            if cue.iter().any(|s| rel_s().contains(s) && !has_stem(&e.s, s)) && !(pet_query && es_pet) { continue; }
            if e.id < 0 { continue; }
            *act.entry(e.id).or_insert(0.0) += ov as f64 * self.eff(e.id).max(1e-6);
        }
        if act.is_empty() { return Vec::new(); }
        let seeds: HashSet<i64> = act.keys().cloned().collect();
        let mut frontier = act.clone();
        for _ in 0..hops {
            let mut nxt: HashMap<i64, f64> = HashMap::new();
            for (&src, &a) in frontier.iter() {
                if a <= 1e-9 { continue; }
                let nbrs = match self.links.get(&src) { Some(m) => m, None => continue };
                let mut top: Vec<(i64, f64)> = nbrs.iter().map(|(a, b)| (*a, *b)).collect();
                top.sort_by(|x, y| (y.1 * self.eff(y.0)).partial_cmp(&(x.1 * self.eff(x.0))).unwrap());
                top.truncate(fan);
                let norm: f64 = top.iter().map(|(_, w)| *w).sum::<f64>().max(1.0);
                for (nid, w) in top {
                    let sig = a * (1.0 - reuptake) * (w / norm) * self.eff(nid).min(1.0);
                    if sig <= 1e-9 { continue; }
                    *nxt.entry(nid).or_insert(0.0) += sig;
                }
            }
            for (nid, s) in nxt.iter() { *act.entry(*nid).or_insert(0.0) += *s; }
            if nxt.is_empty() { break; }
            frontier = nxt;
        }
        let mut ranked: Vec<(i64, f64)> = act.into_iter().collect();
        ranked.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
        ranked.truncate(k);
        ranked.into_iter().filter_map(|(id, a)| by_id.get(&id).map(|&ix| {
            let e = &self.base.episodes[ix];
            Spread { fact: e.t.clone(), value: e.v.clone(), activation: (a * 1e4).round() / 1e4, seed: seeds.contains(&id) }
        })).collect()
    }

    /// "Sleep": merge facts with identical stem-sets (sum strength), prune decayed-away facts.
    pub fn consolidate(&mut self, prune_below: f64) -> (usize, usize, usize) {
        let mut by_key: HashMap<Vec<Arc<str>>, i64> = HashMap::new();
        let mut keep: Vec<usize> = Vec::new();
        let (mut merged, mut pruned) = (0usize, 0usize);
        for i in 0..self.base.episodes.len() {
            let key = self.base.episodes[i].s.clone();
            let id = self.base.episodes[i].id;
            if let Some(&ai) = by_key.get(&key) {
                let e = self.eff(ai) + self.eff(id); self.w.insert(ai, e); self.t.insert(ai, self.tick);
                if let Some(m) = self.links.remove(&id) { for (nid, w) in m { self.link(ai, nid, w); } }
                merged += 1;
            } else { by_key.insert(key, id); keep.push(i); }
        }
        let mut survivors: Vec<usize> = Vec::new();
        for &i in &keep {
            let (id, is_self) = { let e = &self.base.episodes[i]; (e.id, e.self_flag) };
            if self.eff(id) >= prune_below || is_self || self.pinned.contains(&id) { survivors.push(i); }
            else { pruned += 1; self.w.remove(&id); self.t.remove(&id); self.links.remove(&id); }
        }
        let survivor_set: HashSet<usize> = survivors.iter().cloned().collect();
        let mut new_eps = Vec::new();
        for i in 0..self.base.episodes.len() { if survivor_set.contains(&i) { new_eps.push(self.base.episodes[i].clone()); } }
        self.base.episodes = new_eps;
        // drop any link edge that points at a merged-away or pruned id — both endpoints must still be
        // alive — so spreading recall never chases a dangling neighbor (incoming refs, not just outgoing).
        let alive: HashSet<i64> = self.base.episodes.iter().map(|e| e.id).collect();
        self.links.retain(|id, _| alive.contains(id));
        for nbrs in self.links.values_mut() { nbrs.retain(|nid, _| alive.contains(nid)); }
        self.base.invalidate_index();
        (merged, pruned, self.base.episodes.len())
    }
}
